372 add spade normalisation option for diffusion models

virginiafdez opened the pull request:
Added SPADE-LDM code:

- Modification of diffusion_model_unet to allow for SPADE normalisation to be set up as an option
- Modification of autoencoder_kl to allow for SPADE normalisation to be set up as an option
- Modification of inferer and latent inferer to allow for label to be passed through forward when SPADE is active
- Addition of tests to: test_spade_diffusion
- Creation of tutorial for 2D using OASIS subset of images.

Even though I implemented tests, we should check very thoroughly that this works before merging, especially since the presence of SPADE norm needs for labels to be passed to the forward method, and ANY call of forward without a label if SPADE Is on will end up in error. In the same fashion, we should ensure that ANY call on forward when SPADE is not on is not disrupted (code doesn't error out because of a label missing).

marksgraham commented on tutorials/generative/2d_ldm/2d_ldm_tutorial.ipynb:
```diff
@@ -1543,7 +1543,7 @@
    "nbconvert_exporter": "python",
    "pygments_lexer": "ipython3",
-   "version": "3.10.6"
+   "version": "3.9.16"
```
no real change to this file, please remove from the PR

marksgraham commented on tutorials/generative/2d_spade_gan/2d_spade_vae.ipynb:
```diff
@@ -2397,7 +2397,7 @@
    "nbconvert_exporter": "python",
    "pygments_lexer": "ipython3",
-   "version": "3.8.13"
+   "version": "3.9.16"
```
please remove this and the corresponding .py file from the PR, too

marksgraham commented on tutorials/generative/2d_spade_ldm/2d_spade_ldm.py:
```diff
@@ -0,0 +1,450 @@
+# limitations under the License.
+# -
+
+# # SPADE LDM
```
please put a short intro to the notebook here, including what it will walk through and linking to relevant publications, too

marksgraham commented on tutorials/generative/2d_spade_ldm/2d_spade_ldm.py:
```diff
@@ -0,0 +1,450 @@
+zip_obj = zipfile.ZipFile(os.path.join(root_dir, "data.zip"), "r")
+zip_obj.extractall(root_dir)
+images_T1 = root_dir / "OASIS_SMALL-SUBSET/T1"
+images_FLAIR = root_dir / "OASIS_SMALL-SUBSET/FLAIR"
```
these FLAIR images don't get used anywhere, maybe best to remove references to them?

marksgraham reviewed (changes requested):
Tutorial looks good! 

I think we should maybe spin the spade DDPM and spade autoencoder out into their own models to keep the network code cleaner - can discuss together!

virginiafdez commented:
@marksgraham  I just pushed the changes:

- I removed all the changes from autoencoderkl and diffusion_model_unet and created separate spade_autoencoderkl and spade_diffusion_model_unet scripts.  The logic now is that, obviously, objects from these classes have SPADE and you need to pass label_nc argument. 
- Based on that, I modified the inferer.py to accept optional seg, checking if any (AE or DM) are instances of SPADEAutoencoderKL or SPADEDiffusionModelUNet
- I changed the tests for the inferer to add SPADE cases
- I added tests for spade_autoencoderkl and spade_diffusion_model_unet
- I modified the tutorial to incorporate this name changes.
